### ISSUE TITLE
The dart2js transformer runs out of memory compiling multiple entrypoints

### ISSUE THREAD

**Issue by DartBot** [dart-lang/pub#701]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="96" height="96"hspace="10"></img></a> **Issue by [munificent](https://github.com/munificent)**
_Originally opened as dart-lang/sdk#14730_

----

Siggy (and others) can consistently repro an issue where pub build will crash with out of memory. To do so, just duplicate the index.html file in web/ and then add it as an entrypoint:

transformers:
- polymer:
&nbsp;&nbsp;&nbsp;&nbsp;entry_points: [web/index.html, web/index2.html]

The issue seems to be that dart2js does \*some\* of its work asynchronously, which means we get multiple compiles going at least a little bit in parallel, and there isn't enough room for it.

We should investigate this more and have a clearer picture of what's going on. We can work around it by making the transformer force itself to be sequential, but a cleaner fix would be better.

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Added this to the [Later](../milestones/Later) milestone.*

**Comment by DartBot** [comment]
<a href="https://github.com/sigmundch"><img src="https://avatars.githubusercontent.com/u/2049220?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [sigmundch](https://github.com/sigmundch)**

----

Issue dart-lang/sdk#15054 has been merged into this issue.

**Comment by DartBot** [comment]
<a href="https://github.com/zoechi"><img src="https://avatars.githubusercontent.com/u/405837?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [zoechi](https://github.com/zoechi)**

----

**Attachment:**
[pub_build_out_of_memory.txt](https://storage.googleapis.com/google-code-attachments/dart/issue-14730/comment-3/pub_build_out_of_memory.txt) (305.51 KB)

**Comment by DartBot** [comment]
<a href="https://github.com/zoechi"><img src="https://avatars.githubusercontent.com/u/405837?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [zoechi](https://github.com/zoechi)**

----

**Attachment:**
[build.zip](https://storage.googleapis.com/google-code-attachments/dart/issue-14730/comment-4/build.zip) (103.54 KB)

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

On my mac laptop, I was able to get it to build completely, but it did peak at around 600MB usage.

If I disable dart2js, I see it still eating about 260MB even before it prints "Building polymer_elements_examples....", so it looks like we're using a decent amount of memory just in barback. This is for a build that has 474 source assets, which is not a huge number.

**Comment by DartBot** [comment]
<a href="https://github.com/zoechi"><img src="https://avatars.githubusercontent.com/u/405837?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [zoechi](https://github.com/zoechi)**

----

For me it still fails with 'out of memory' on a 12GB Debian linux system with or without --mode=debug

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Removed [Priority-Unassigned](../labels/Priority-Unassigned) label.*
*Added [Priority-High](../labels/Priority-High) label.*

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Marked this as blocking dart-lang/sdk#15829.*

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Marked this as being blocked by dart-lang/sdk#8355.*

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

*Added [DevStory](../labels/DevStory) label.*

**Comment by DartBot** [comment]
<a href="https://github.com/munificent"><img src="https://avatars.githubusercontent.com/u/46275?v=3" align="left" width="48" height="48"hspace="10"></img></a> **Comment by [munificent](https://github.com/munificent)**

----

With latest code, it seems like we're doing OK. I just ran the following test:

- A package containing 50 copies of the TodoMVC web directory.
- 50 polymer entrypoints
- 1,101 source files (1.93 MB)
- 417 output files (72.0 MB)

Although it took a while, it completed successfully on both my Mac and Windows machine. Memory usage quickly went to about 450 MB and then hovered around there, gradually creeping up to around 500 MB before completing.
______
*Added [AssumedStale](../labels/AssumedStale) label.*
